In [45]:
CLEAN_DATA_PATH = './data/Cleaned_Data/cleanData.pkl'
INTERESTED_VARIABLES = ['Rating','Revenue','Average Salary','State','Company_Age_years','python','R','Job_Experience',
'spark','AWS','excel','Job_Specification','Competitors_Number','SalaryPerHour','Employer Provided Salary', 'Type of ownership']

In [46]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn

%matplotlib inline

data = pd.read_pickle(CLEAN_DATA_PATH)


In [47]:

copyData = data[INTERESTED_VARIABLES]


In [48]:
Data_with_dummies = pd.get_dummies(copyData)
Data_with_dummies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 742 entries, 0 to 955
Data columns (total 84 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Rating                                            742 non-null    float64
 1   Average Salary                                    742 non-null    float64
 2   Company_Age_years                                 742 non-null    int64  
 3   python                                            742 non-null    int64  
 4   R                                                 742 non-null    int64  
 5   spark                                             742 non-null    int64  
 6   AWS                                               742 non-null    int64  
 7   excel                                             742 non-null    int64  
 8   Competitors_Number                                742 non-null    int64  
 9   SalaryPerHour        

In [49]:
Data_with_dummies_new = Data_with_dummies.copy()#drop(['Revenue_-1', 'State_AL', 'Job_Specification_Research Scientist'], axis=1)
Data_with_dummies_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 742 entries, 0 to 955
Data columns (total 84 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Rating                                            742 non-null    float64
 1   Average Salary                                    742 non-null    float64
 2   Company_Age_years                                 742 non-null    int64  
 3   python                                            742 non-null    int64  
 4   R                                                 742 non-null    int64  
 5   spark                                             742 non-null    int64  
 6   AWS                                               742 non-null    int64  
 7   excel                                             742 non-null    int64  
 8   Competitors_Number                                742 non-null    int64  
 9   SalaryPerHour        

# Build Models

In [50]:
from sklearn.model_selection import train_test_split

y = Data_with_dummies_new['Average Salary'].values
X = Data_with_dummies_new.drop('Average Salary', axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


### Model 1 OLS

In [51]:
# first build model use OLS from statemodel
import statsmodels.api as stat

X_sm = X = stat.add_constant(X)
model = stat.OLS(y,X_sm)
model.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.685
Model:                            OLS   Adj. R-squared:                  0.648
Method:                 Least Squares   F-statistic:                     18.74
Date:                Thu, 11 Nov 2021   Prob (F-statistic):          7.68e-122
Time:                        14:56:03   Log-Likelihood:                -8498.2
No. Observations:                 742   AIC:                         1.715e+04
Df Residuals:                     664   BIC:                         1.751e+04
Df Model:                          77                                         
Covariance Type:            nonrobust                                         
====================================================================================================================
                                                       coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------------------
const                                             5.202e+04   5624.139      9.250      0.000     4.1e+04    6.31e+04
Rating                                            2418.8161   1443.787      1.675      0.094    -416.122    5253.754
Company_Age_years                                   20.0543     25.793      0.778      0.437     -30.591      70.699
python                                            7836.0658   2295.563      3.414      0.001    3328.630    1.23e+04
R                                                -1.258e+04   2.01e+04     -0.627      0.531    -5.2e+04    2.68e+04
spark                                             2623.3158   2716.835      0.966      0.335   -2711.307    7957.939
AWS                                                -58.8216   2463.320     -0.024      0.981   -4895.657    4778.014
excel                                               -1.9903   2028.231     -0.001      0.999   -3984.509    3980.528
Competitors_Number                                1950.5235    761.356      2.562      0.011     455.568    3445.479
SalaryPerHour                                    -6.738e+04   7706.020     -8.744      0.000   -8.25e+04   -5.22e+04
Employer Provided Salary                          2.957e+04   7911.620      3.737      0.000     1.4e+04    4.51e+04
Revenue_$1 to $2 billion (USD)                    4728.0784   3687.427      1.282      0.200   -2512.344     1.2e+04
Revenue_$1 to $5 million (USD)                   -8980.7847   8892.371     -1.010      0.313   -2.64e+04    8479.769
Revenue_$10 to $25 million (USD)                 -4552.8093   4662.998     -0.976      0.329   -1.37e+04    4603.189
Revenue_$10+ billion (USD)                        8531.9281   3734.571      2.285      0.023    1198.938    1.59e+04
Revenue_$100 to $500 million (USD)                -959.3862   3295.647     -0.291      0.771   -7430.531    5511.758
Revenue_$2 to $5 billion (USD)                   -3833.1077   4433.135     -0.865      0.388   -1.25e+04    4871.544
Revenue_$25 to $50 million (USD)                  -187.2390   4535.356     -0.041      0.967   -9092.606    8718.128
Revenue_$5 to $10 billion (USD)                     67.1655   6033.909      0.011      0.991   -1.18e+04    1.19e+04
Revenue_$5 to $10 million (USD)                    2.62e+04   6361.961      4.118      0.000    1.37e+04    3.87e+04
Revenue_$50 to $100 million (USD)                 1.177e+04   4056.372      2.901      0.004    3802.780    1.97e+04
Revenue_$500 million to $1 billion (USD)          -315.3333   4326.032     -0.073      0.942   -8809.684    8179.017
Revenue_-1                                        1.975e+04   1.19e+04      1.663      0.097   -3564.002    4.31e+04
Revenue_Less than $1 million (USD)       

## Linear Regression Model

In [52]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso

linearModel = LinearRegression()
linearModel.fit(X_train, y_train)

print("Training set score: {:.2f}".format(linearModel.score(X_train, y_train))) 
print("Test set score: {:.2f}".format(linearModel.score(X_test, y_test)))

Training set score: 0.69
Test set score: 0.65


**Ridge**

In [53]:
ridgeModel = Ridge(alpha=0.05)
ridgeModel.fit(X_train, y_train)

print("Training set score: {:.2f}".format(ridgeModel.score(X_train, y_train)))
print("Test set score: {:.2f}".format(ridgeModel.score(X_test, y_test)))


Training set score: 0.69
Test set score: 0.65


**Lasso**

In [54]:
lassoModel = Lasso(max_iter=100000, alpha=0.01)
lassoModel.fit(X_train, y_train)

print("Training set score: {:.2f}".format(lassoModel.score(X_train, y_train)))
print("Test set score: {:.2f}".format(lassoModel.score(X_test, y_test)))

Training set score: 0.69
Test set score: 0.65


## RandomForestRegressor Model

In [107]:
from sklearn.ensemble import RandomForestRegressor
import numpy as np
rf = RandomForestRegressor()

rf.fit(X_train, y_train)


print("Training set score: {:.2f}".format(rf.score(X_train, y_train)))
print("Test set score: {:.2f}".format(rf.score(X_test, y_test)))


Training set score: 0.95
Test set score: 0.77


Evaluatiom Model

In [105]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import GridSearchCV
import numpy as np
rf = RandomForestRegressor()
kfold = KFold(n_splits=5, random_state=0, shuffle=True)
sss = cross_val_score(rf, X_train, y_train, cv=kfold)

rf.fit(X_train, y_train)
p = rf.predict(X_test)
print("Training KFold acc: {}".format(np.mean(sss)))

# Parameter Tuning Using Grid Search
# absolute_error, squared_error
pr = {'n_estimators':range(10,300,10), 'criterion':('mse','mae'), 'max_features':('auto','sqrt','log2')}
gs = GridSearchCV(rf,pr,cv=kfold)
gs.fit(X_train,y_train)

gs.best_score_
gs.best_estimator_
# print("Training set score: {:.2f}".format(rf.score(X_train, y_train)))
# print("Test set score: {:.2f}".format(rf.score(X_test, y_test)))


Training KFold acc: 0.7131421571743529


C:\Users\Alien\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\ensemble\_forest.py:387: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warn(
C:\Users\Alien\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\ensemble\_forest.py:387: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warn(
C:\Users\Alien\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\ensemble\_forest.py:387: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warn(
C:\Users\Alien\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\ensemble\_forest.py:387: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which

RandomForestRegressor(criterion='mse', max_features='sqrt', n_estimators=230)

In [106]:
print("Training set score: {:.2f}".format(gs.best_estimator_.score(X_train, y_train)))
print("Test set score: {:.2f}".format(gs.best_estimator_.score(X_test, y_test)))

Training set score: 0.96
Test set score: 0.80


### Notice that the best model is **RandomForestRegressor** that gives training score 96 % and testing score 80 % 